### PR TITLE
Track AI topic approval by workflow owner

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
This PR adds a boolean field that tracks if the user selected the AI topic. The UI uses this field to determine if the AI bubble should be displayed. If the topic has been user-selected, then it's not displayed publicly. Otherwise, it's displayed.

See https://github.com/dockstore/dockstore-ui2/pull/1997 for UI changes.

**Review Instructions**
See instructions in https://github.com/dockstore/dockstore-ui2/pull/1997

**Issue**
https://github.com/dockstore/dockstore/issues/5866
https://ucsc-cgl.atlassian.net/browse/DOCK-2521

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 

- [x] Security and Privacy assessed

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java:437]: Hmmmm, is this the general update code?
What if the user just means to update the forum url for example and did not mean to approve the current ai topic?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:571]: Ditto, for example if the user meant just to update the defaultWorkflowPath, would the approval be set to true accidentally?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java:437]: Hmm good point. My idea was that if the user selected AI topic then it automatically means that they have reviewed and approved it, so it's kind of redundant to pass along `approvedAITopic`. But in your scenario, it can lead to accidental approval.

Kind of makes me want to separate out the topic change updates to its own endpoint :thinking:

Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java:437]: That could work too, but I don't think it is (always) redundant to pass along approvedAITopic, I think you have a few scenarios

a) User means to change something else entirely (selection stays the same as before, approved topic doesn't change) 
b) User means to approve the AI topic (selection stays, boolean changes to true)
c) User means to change to .dockstore.yml topic (selection changes, boolean is now redundant)
d) User changes back to AI topic (selection changes, boolean goes back to true and is not redundant)

In other words, the boolean is redundant if the topic is not AI generated.

**Review by coverbeck — APPROVED** [review]
Pending addressing Denis' comments.

**Review by denis-yuen — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java:665]: I wonder, if the user has already selected the AI topic and implicitly approved it, should we reject the attempt to change it, or something like that?  This would help to prevent us from accidentally overriding the user's previous action.  Of course, we would tweak the topic updater support app to perform the same check, this would be a backup.

Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java:665]: Perhaps I should perform a check to see if the new AI topic is different and only reset the `approvedAITopic` field to false if it is? That would prevent us from accidentally resetting the user's approval if the topic has not changed.

If the AI topic has actually changed, I think it will need another approval from the user because the content has changed

Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java:665]: Hmm, some random obs:

* Per my understanding of ChatGPT/etc, the generated topic will vary each time, even if we submit the same input.
* It'd be a bummer for the user to have to keep selecting/approving generated topics, even after trivial changes to the entry.
* We do want the user to know about and use a newly-generated topic, if the entry has changed enough that the previously-generated topic is obsolete.

Maybe before generation, we could ask ChatGPT if the existing topic still applies?  Or, after generating a new topic, ask ChatGPT if the new and previous topics mean the same thing, and not update if they are semantically the same?

Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java:665]: (the above suggests that the we do any necessary checks in the topic generator, and that the webservice logic you've implemented is good.)

Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java]: > It'd be a bummer for the user to have to keep selecting/approving generated topics, even after trivial changes to the entry.

Agreed, I added an additional check in the webservice so it only resets the approval if the topic has changed

> We do want the user to know about and use a newly-generated topic, if the entry has changed enough that the previously-generated topic is obsolete.

I wonder if we should even be re-generating AI topics if they've already approved of it. Maybe we allow them to choose if they want it regenerated? That would probably require some extra work though depending if we want instant re-generation and the 1.16 priority is likely generating topics for old and new entries without any topics.

> Maybe before generation, we could ask ChatGPT if the existing topic still applies? Or, after generating a new topic, ask ChatGPT if the new and previous topics mean the same thing, and not update if they are semantically the same?

Good idea, I created a ticket for that https://ucsc-cgl.atlassian.net/browse/SEAB-6590

**Review by svonworl — APPROVED** [review]

**Pull request merged**
